src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1213 @@
+        let bhh_2 = format!("{:02x}", bhh_bytes[30]);
+        let p = Path::new(dir_path)
+                    .join(bhh_1)
+                    .join(bhh_2);
+        p
+    }
+
+    pub fn block_path(dir_path: &String, bhh: &BlockHeaderHash) -> PathBuf {
+        // it looks awkward, but it's waaaay faster than just doing to_hex()
+        let bhh_bytes = bhh.as_bytes();
+        let bhh_name = format!("{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
+                              bhh_bytes[0],     bhh_bytes[1],       bhh_bytes[2],       bhh_bytes[3],
+                              bhh_bytes[4],     bhh_bytes[5],       bhh_bytes[6],       bhh_bytes[7],
+                              bhh_bytes[8],     bhh_bytes[9],       bhh_bytes[10],      bhh_bytes[11],
+                              bhh_bytes[12],    bhh_bytes[13],      bhh_bytes[14],      bhh_bytes[15],
+                              bhh_bytes[16],    bhh_bytes[17],      bhh_bytes[18],      bhh_bytes[19],
+                              bhh_bytes[20],    bhh_bytes[21],      bhh_bytes[22],      bhh_bytes[23],
+                              bhh_bytes[24],    bhh_bytes[25],      bhh_bytes[26],      bhh_bytes[27],
+                              bhh_bytes[28],    bhh_bytes[29],      bhh_bytes[30],      bhh_bytes[31]);
+
+        TrieFileStorage::block_dir(dir_path, bhh).join(bhh_name)
+    }
+
+    pub fn block_path_tmp(dir_path: &String, bhh: &BlockHeaderHash) -> PathBuf {
+        // it looks awkward, but it's waaaay faster than just doing to_hex()
+        let bhh_bytes = bhh.as_bytes();
+        let bhh_name = format!("{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}.tmp",
+                              bhh_bytes[0],     bhh_bytes[1],       bhh_bytes[2],       bhh_bytes[3],
+                              bhh_bytes[4],     bhh_bytes[5],       bhh_bytes[6],       bhh_bytes[7],
+                              bhh_bytes[8],     bhh_bytes[9],       bhh_bytes[10],      bhh_bytes[11],
+                              bhh_bytes[12],    bhh_bytes[13],      bhh_bytes[14],      bhh_bytes[15],
+                              bhh_bytes[16],    bhh_bytes[17],      bhh_bytes[18],      bhh_bytes[19],
+                              bhh_bytes[20],    bhh_bytes[21],      bhh_bytes[22],      bhh_bytes[23],
+                              bhh_bytes[24],    bhh_bytes[25],      bhh_bytes[26],      bhh_bytes[27],
+                              bhh_bytes[28],    bhh_bytes[29],      bhh_bytes[30],      bhh_bytes[31]);
+
+        TrieFileStorage::block_dir(dir_path, bhh).join(bhh_name)
+    }
+
+    fn cached_block_path(&mut self, bhh: &BlockHeaderHash) -> PathBuf {
+        let (p, miss) = match self.block_path_cache.get(bhh) {
```

Comment:
what's the upside of caching block_paths? what is the expensive thing about turning a BlockHeaderHash (encoded as bytes) into just a hex string representation of those bytes? is it object allocs?

Reply:
Granted, I haven't verified this in `perf` with `--release`, but at least in debug builds, string formatting is surprisingly expensive.  This avoids the need for an expensive `format!` call.